### ISSUE TITLE
Implement movement based on unicode segmentation

### ISSUE THREAD

**Issue by kirawi** [helix-editor/helix#11423]
I think it would be good to have better support for other languages where the concept of sentence and word boundaries vary. Implementing word-based movements equivalent to the current ones would be helpful. For example, it is expected that in あいうアイウ, あいう and アイウ are treated as separate words. Implementing this would likely require implementing the text segmentation algorithm in Helix since `unicode-segmentation` does not offer a way to implement it on non-contiguous text.

**Comment by RossSmyth** [comment]
There is also the ICU segmenter, but it may be more than desired. It also lacks the ability to load non-contiguous text.
https://docs.rs/icu_segmenter/latest/icu_segmenter/

**Comment by kirawi** [comment]
We could also fork unicode-width and add support for ropey directly

**Comment by RossSmyth** [comment]
I was curious to see how difficult it would be to do, so I poked at the ICU word segmenter. It turned out to be mostly straight-forward to make it compile.
https://github.com/unicode-org/icu4x/compare/main...RossSmyth:icu4x:Ropey

I did not test anything but it does build.

With the above it should work something like (all untested):

```rust
use icu::segmenter::WordSegmenter;
use ropey::Rope;

let segmenter = WordSegmenter::new_auto();
let rope = Rope::from_str("Hello World");

let breakpoints: Vec<usize> =
    segmenter.segment_str(rope.chars()).collect();
assert_eq!(&breakpoints, &[0, 5, 6, 11]);

let breakpoints: Vec<usize> =
    segmenter.segment_str(rope.chars_at(6)).collect();
assert_eq!(&breakpoints, &[0, 5]);
```

**Comment by kirawi** [comment]
I don't think we want to pull in icu because of how many dependencies it brings in.

**Comment by kirawi** [comment]
@pascalkuthe @the-mikedavis I took a look at creating a fork, and it looks to be feasible to write a trait instead of just a string for segmentation. Do we care about making it generic, or would it be fine to just forego that idea in favor of switching from `str` to `RopeSlice`?

**Comment by imnotlxy** [comment]
I think introducing ICU is the best choice. ICU has decent Trie implementation and a highly-compressed compiled dictionary file. It's a lot of code to implement it again. A toy copy of this dictionary-based segmenter: https://gitlab.gnome.org/GNOME/pango/-/merge_requests/905. Seems that nobody likes it, because the algorithm is less optimized and the dictionary isn't as compact.

A practical idea is to isolate the ICU dependencies in a pluggable `.so` file, and to provide an option in `config.toml`.
